xdsl/parser.py

```diff
@@ -1,1335 +1,1942 @@
-                "false", skip_white_space=skip_white_space) is not None:
-            return IntegerAttr.from_int_and_width(0, 1)
-
-    def parse_optional_xdsl_builtin_attribute(self,
-                                              skip_white_space: bool = True
-                                              ) -> Attribute | None:
-        # Shorthand for StringAttr
-        string_lit = self.parse_optional_str_literal(
-            skip_white_space=skip_white_space)
-        if string_lit is not None:
-            return StringAttr.from_str(string_lit)
-
-        # Shorthand for FloatAttr
-        float_lit = self.parse_optional_float_literal()
-        if float_lit is not None:
-            if self.parse_optional_char(":"):
-                typ = self.parse_attribute()
-            else:
-                typ = Float32Type()
-            return FloatAttr.from_value(float_lit, typ)
-
-        # Shorthand for boolean literals (IntegerAttr of width 1)
-        if (bool_attr := self.parse_optional_boolean_attribute(
-                skip_white_space=skip_white_space)):
-            return bool_attr
-
-        # Shorthand for IntegerAttr
-        integer_lit = self.parse_optional_int_literal()
-        if integer_lit is not None:
-            if self.parse_optional_char(":"):
-                typ = self.parse_attribute()
-            else:
-                typ = IntegerType.from_width(64)
-            return IntegerAttr.from_params(integer_lit, typ)
-
-        # Shorthand for ArrayAttr
-        parse_bracket = self.parse_optional_char("[")
-        if parse_bracket:
-            array = self.parse_list(self.parse_optional_attribute)
-            self.parse_char("]")
-            return ArrayAttr.from_list(array)
-
-        # Shorthand for DictionaryAttr
-        parse_bracket = self.parse_optional_char("{")
-        if parse_bracket:
-            dictionary = self.parse_dictionary(self.parse_optional_str_literal,
-                                               self.parse_optional_attribute)
-            self.parse_char("}")
-            return DictionaryAttr.from_dict(dictionary)
-
-        # Shorthand for FlatSymbolRefAttr
-        parse_at = self.parse_optional_char("@")
-        if parse_at:
-            symbol_name = self.parse_alpha_num(skip_white_space=False)
-            return FlatSymbolRefAttr.from_str(symbol_name)
-
-        def parse_integer_type():
-            self.parse_char("!", skip_white_space=skip_white_space)
-            return self.parse_mlir_integer_type(
-                skip_white_space=skip_white_space)
-
-        if int_type := self.try_parse(parse_integer_type):
-            return int_type
+        builtin_parsers: dict[str, Callable[[], ParametrizedAttribute]] = {
+            "vector": self.must_parse_vector_attrs,
+            "memref": self.must_parse_memref_attrs,
+            "tensor": self.must_parse_tensor_attrs,
+            "complex": self.must_parse_complex_attrs,
+            "tuple": unimplemented,
```

Comment:
I don't like this connection of the parser with the defined attributes, but this is probably a problem for a different day...

Reply:
As far as I understand xDSL/MLIR, the parser must implement the builtin dialect parsing. I could be wrong though. Every time I talked to Mathieu about that he told me how it should all change in the future, so I was under the impression that this was a somewhat temporary thing anyway.

Reply:
Yes, agreed. The design sucks, but we only have that atm...

Reply:
Yeah, for now, we should keep this since MLIR does the same, and hopefully, in the future, we could make this support optional, or extensible at least.